src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Helpers.cs

```diff
@@ -468,5 +468,119 @@ internal static void ThrowIfPasswordContainsNullCharacter(string? password)
 
         internal static bool IsSlhDsaOid(string? oid) =>
             SlhDsaAlgorithm.GetAlgorithmFromOid(oid) is not null;
+
+        public delegate void SlhDsaPrehashActionCallback<T, TState>(Span<T> span, TState arg)
+            where TState : allows ref struct;
+
+        public delegate TResult SlhDsaPrehashFuncCallback<T, TState, TResult>(Span<T> span, TState arg)
+            where TState : allows ref struct;
+
+        internal static void SlhDsaPrehash<TState>(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> context,
+            HashAlgorithmName preHashAlgorithm,
+            TState state,
+            SlhDsaPrehashActionCallback<byte, TState> callback)
+            where TState : allows ref struct
+        {
+            SlhDsaPrehash(
+                hash,
+                context,
+                preHashAlgorithm,
+                StackTuple.Create(state, callback),
+                static (message, myState) =>
+                {
+                    myState.Item2(message, myState.Item1);
+                    return true;
+                });
+        }
+
+        internal static TResult SlhDsaPrehash<TState, TResult>(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> context,
+            HashAlgorithmName preHashAlgorithm,
+            TState state,
+            SlhDsaPrehashFuncCallback<byte, TState, TResult> callback)
+            where TState : allows ref struct
+        {
+            string? oid = Helpers.GetOidFromHashAlgorithm(preHashAlgorithm);
+
+            // TODO should we throw instead?
+            Debug.Assert(oid != null, $"Hash algorithm {preHashAlgorithm} must have already been validated.");
+
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.WriteObjectIdentifier(oid);
```

Comment:
One of these days we'll get around to exposing AsnEncoder to get the non-allocating version of turning the OID into bytes :smile:.